dali/operators/audio/mfcc/mfcc.cc

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/mfcc/mfcc.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/signal/dct/dct_cpu.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/common/for_axis.h"
+#include "dali/pipeline/data/views.h"
+
+
+#define MFCC_SUPPORTED_TYPES (float)
+#define MFCC_SUPPORTED_NDIMS (2, 3, 4)
+
+static constexpr int kNumInputs = 1;
+static constexpr int kNumOutputs = 1;
+
+namespace dali {
+
+namespace detail {
+
+template <typename T, int Dims>
+void ApplyLifter(const kernels::OutTensorCPU<T, Dims> &inout, int axis, const T* lifter_coeffs) {
+  assert(axis >= 0 && axis < Dims);
+  assert(lifter_coeffs != nullptr);
+  auto* data = inout.data;
+  auto shape = inout.shape;
+  auto strides = kernels::GetStrides(shape);
+  kernels::ForAxis(
```

Comment:
I feel like this can be optimized a bit, don't know if it's worth the effort, but will probably yield better memory access pattern.
You can have a variant of ForAxis for a case like Dims=5, axis=2 and do soemthing like:
```
for (x0 in Dim0)
  for (x1 in Dim1):
    // we're on our target axis, now we will use the same lifter_coefficient for neighbour
    // elements, so instead of invoking this lambda and iterating in Dim2 by jumping around
    // the data, calculate it in groups for coefficient 0, than 1, etc
   for (x2 in Dim2)
     multiply all Dim3 * Dim4 elements by lifter[x2]
```

Reply:
I agree that in certain configurations this could be optimized for a better access pattern. However, ForAxis is rather a general utility and this would be an optimization that is specific for this particular case (because we can reuse the lifter coefficient). We could write something custom for a certain layout but I think it would be best to keep simplicity/generality here unless we know that there is a real performance problem here.

Reply:
Ack